clib/hashcons.mli

```diff
@@ -30,30 +30,25 @@ module type HashconsedType =
     type t
     (** Type of objects to hashcons. *)
 
-    val hashcons : t -> t
-    (** The actual hashconsing function, using its first argument to recursively
-        hashcons substructures. It should be compatible with [eq], that is
-        [eq x (hashcons f x) = true]. *)
+    val hashcons : t -> int * t
+    (** The actual hashconsing function. It should be compatible with [eq], that is
+        [eq x (snd @@ hashcons f x) = true]. It also returns the hash. *)
 
     val eq : t -> t -> bool
     (** A comparison function. It is allowed to use physical equality
         on the sub-terms hashconsed by the [hashcons] function, but it should be
-        insensible to shallow copy of the compared object. *)
-
-    val hash : t -> int
-    (** A hash function passed to the underlying hashtable structure. [hash]
-        should be compatible with [eq], i.e. if [eq x y = true] then
-        [hash x = hash y]. *)
+        insensible to shallow copy of the compared object.
+        It should be compatible with the hash returned by [hashcons], ie
+        [eq x y] implies [fst @@ hashcons x = fst @@ hashcons y].
+    *)
   end
 
 module type HashconsedRecType = sig
   type t
 
-  val hashcons : (t -> t) -> t -> t
+  val hashcons : (t -> int * t) -> t -> int * t
```

Comment:
For readability, I think you could introduce a `type 'a hcons = 'a -> int * 'a` and use this in every API that is tweaked by this PR.

Reply:
changed